File: infrastructures/ddd/typescript/src/domain/events/infra/BaseDomainEvent.ts
import { randomUUID } from 'crypto';

/**
 * Base class for all domain events.
 * Domain events:
 * - Live in /domain/events/
 * - Suffix with "Event"
 * - Follow pattern: {Object}{Action}Event
 * - Are immutable (readonly properties)
 * - Include eventId, timestamp, and correlationId
 * - Used for subdomain-to-subdomain communication ONLY
 * 
 * Domain events are the ONLY way to communicate between subdomains.
 * NO HTTP calls between subdomains.
 */
export abstract class BaseDomainEvent {
  /**
   * Unique identifier for this event
   */
  public readonly EventId: string;

  /**
   * Timestamp when the event was created
   */
  public readonly Timestamp: Date;

  /**
   * Correlation ID for tracking requests across services
   */
  public readonly CorrelationId: string;

  protected constructor(p_correlationId: string) {
    this.EventId = randomUUID();
    this.Timestamp = new Date();
    this.CorrelationId = p_correlationId;
  }
}

File: infrastructures/ddd/python/src/domain/events/infra/base_domain_event.py
"""
Base class for all domain events.
Domain events MUST:
- Suffix with "Event"
- Follow pattern: {Object}{Action}Event
- Be immutable
- Live in /domain/events/
- Include event_id, timestamp, and correlation_id
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Optional
import uuid


@dataclass(frozen=True)
class BaseDomainEvent:
    """
    Base class for all domain events.
    
    Domain events are used for:
    - Asynchronous subdomain-to-subdomain communication
    - ONLY way to communicate between subdomains (NO HTTP calls)
    
    All domain events are immutable (frozen dataclass).
    """
    
    event_id: str
    timestamp: datetime
    correlation_id: str
    
    def __init__(self, p_correlation_id: str):
        """
        Initialize a domain event.
        
        Args:
            p_correlation_id: The correlation ID for tracing
        """
        # Use object.__setattr__ because dataclass is frozen
        object.__setattr__(self, 'event_id', str(uuid.uuid4()))
        object.__setattr__(self, 'timestamp', datetime.utcnow())
        object.__setattr__(self, 'correlation_id', p_correlation_id)
